File: ios/Reguerta/Reguerta/Presentation/ShiftCoverage/ShiftCoverageActions.swift
import Foundation

extension ShiftCoverageViewModel {
    func caseItem(_ id: String) -> ShiftCoverageSnapshot.Case? { snapshot?.cases.first { $0.caseId == id } }
    func memberName(_ id: String) -> String { snapshot?.members?.first { $0.memberId == id }?.displayName ?? id }

    /// Controls are hints from the current projection; the server remains the authority for every command.
    func actions(for item: ShiftCoverageSnapshot.Case) -> [ShiftCoverageCommand.Action] {
        guard let snapshot, !isBusy, pendingCommand == nil, item.writable else { return [] }
        var actions = memberActions(item, snapshot: snapshot)
        if snapshot.isAdmin { actions += adminActions(item, policy: snapshot.policy) }
        if [.open, .offered].contains(item.status), snapshot.isAdmin || item.openedByMe == true {
            actions.append(.cancel)
        }
        return actions
    }

    private func memberActions(
        _ item: ShiftCoverageSnapshot.Case,
        snapshot: ShiftCoverageSnapshot
    ) -> [ShiftCoverageCommand.Action] {
        if item.status == .offered, item.offer?.userId == snapshot.memberId,
           let expires = item.offer?.expiresAtMillis, nowMillis < expires {
            return [.accept, .decline]
        }
        guard item.status == .open, item.selectionPhase == .volunteers,
              let closes = item.volunteerClosesAtMillis, nowMillis < closes, snapshot.eligible else { return [] }
        if item.volunteered { return [.withdrawVolunteer] }
        return item.hasVolunteered == true ? [] : [.volunteer]
    }

    private func adminActions(
        _ item: ShiftCoverageSnapshot.Case,
        policy: ShiftCoverageSnapshot.Policy
    ) -> [ShiftCoverageCommand.Action] {
        switch item.status {
        case .open: selectionActions(item, policy: policy)
        case .offered: nowMillis >= (item.offer?.expiresAtMillis ?? Int64.max) ? [.expire] : []
        case .accepted: nowMillis >= item.scheduledAtMillis ? [.complete, .fail] : [.fail]
        case .cancelled: item.canResumeAdmin == true ? [.resumeAdmin] : []
        case .completed, .failed: []
        }
    }

    private func selectionActions(
        _ item: ShiftCoverageSnapshot.Case,
        policy: ShiftCoverageSnapshot.Policy
    ) -> [ShiftCoverageCommand.Action] {
        switch item.selectionPhase {
        case nil:
            return item.revision == 1 && policy.volunteerWindowMillis != nil ? [.startSelection, .offer] : [.offer]
        case .reserve, .draw: return [.offerNext]
        case .volunteers:
            return nowMillis >= (item.volunteerClosesAtMillis ?? Int64.max) ? [.offerNext] : []
        case .drawRequired:
            guard policy.drawAvailable else { return [] }
            if item.drawCommitted != true { return [.commitDraw] }
            return nowMillis >= (item.drawAvailableAtMillis ?? Int64.max) ? [.revealDraw] : []
        case .adminRequired: return [.offerAdmin]
        }
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/shiftcoverage/ShiftCoverageActions.kt
package com.reguerta.user.presentation.shiftcoverage

import com.reguerta.user.domain.shiftcoverage.ShiftCoverageCommand.Action
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSnapshot
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSnapshot.Phase
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSnapshot.Status

internal fun ShiftCoverageViewModel.actions(item: ShiftCoverageSnapshot.Case): List<Action> {
    val current = state.value
    val snapshot = current.snapshot ?: return emptyList()
    if (current.isBusy || current.pendingCommand != null || !item.writable) return emptyList()
    val actions = mutableListOf<Action>()
    val now = nowMillis
    if (item.status == Status.offered && item.offer?.userId == snapshot.memberId && now < item.offer.expiresAtMillis) {
        actions += listOf(Action.accept, Action.decline)
    }
    if (item.status == Status.open && item.selectionPhase == Phase.volunteers &&
        now < (item.volunteerClosesAtMillis ?: Long.MIN_VALUE) && snapshot.eligible
    ) {
        if (item.volunteered) actions += Action.withdrawVolunteer
        else if (!item.hasVolunteered) actions += Action.volunteer
    }
    if (snapshot.isAdmin) {
        when (item.status) {
            Status.open -> when (item.selectionPhase) {
                null -> {
                    if (item.revision == 1L && snapshot.policy.volunteerWindowMillis != null) actions += Action.startSelection
                    actions += Action.offer
                }
                Phase.reserve, Phase.draw -> actions += Action.offerNext
                Phase.volunteers -> if (now >= (item.volunteerClosesAtMillis ?: Long.MAX_VALUE)) actions += Action.offerNext
                Phase.drawRequired -> if (snapshot.policy.drawAvailable) {
                    if (!item.drawCommitted) actions += Action.commitDraw
                    else if (now >= (item.drawAvailableAtMillis ?: Long.MAX_VALUE)) actions += Action.revealDraw
                }
                Phase.adminRequired -> actions += Action.offerAdmin
            }
            Status.offered -> if (now >= (item.offer?.expiresAtMillis ?: Long.MAX_VALUE)) actions += Action.expire
            Status.accepted -> {
                if (now >= item.scheduledAtMillis) actions += Action.complete
                actions += Action.fail
            }
            Status.cancelled -> if (item.canResumeAdmin) actions += Action.resumeAdmin
            Status.completed, Status.failed -> Unit
        }
    }
    if (item.status in listOf(Status.open, Status.offered) && (snapshot.isAdmin || item.openedByMe)) actions += Action.cancel
    return actions
}

internal fun ShiftCoverageViewModel.memberName(id: String): String =
    state.value.snapshot?.members?.firstOrNull { it.memberId == id }?.displayName ?: id
